### PR TITLE
[CIS-2243, CIS-2244] Increase Safety by making Log Out and Disconnect async operations

### PR DESCRIPTION
**Pull request by polqf**
### 🔗 Issue Links

https://stream-io.atlassian.net/browse/CIS-2243
https://stream-io.atlassian.net/browse/CIS-2244
https://github.com/GetStream/stream-chat-swift/issues/2381

### 🎯 Goal

Avoid race conditions or crashes originated after calling log out

### 📝 Summary

`logOut` does not finish all the "clean up" synchronously, which can potentially create issues when interacting with the SDK just after calling it.

### 🛠 Implementation

Synchronous `logOut` and `disconnect` calls are deprecated in favour of its asynchronous version

### 🧪 Manual Testing Notes

_Explain how this change can be tested manually, if applicable._

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### CODE REVIEW

**Review by martinmitrevski — COMMENTED** [review]
LGTM ✅ We need to update the changelog, check if docs need to be updated and write some tests for the async logout (especially the scenario that previously crashed).

File: Sources/StreamChat/Controllers/ConnectionController/ConnectionController.swift [hunk 1]
```diff
@@ -103,20 +103,12 @@ public extension ChatConnectionController {
     /// Disconnects the chat client the controller represents from the chat servers.
     /// No further updates from the servers are received.
     func disconnect() {
-        chatClientUpdater.disconnect(source: .userInitiated) {
+        connectionRepository.disconnect(source: .userInitiated) {
```
Comment (nuno-vieira) [inline Sources/StreamChat/Controllers/ConnectionController/ConnectionController.swift:106]: Shouldn't this disconnect have a completion block as well?

File: Sources/StreamChat/ChatClient.swift [hunk 2]
```diff
@@ -367,27 +364,52 @@ public class ChatClient {
...
     /// are received.
+    @available(*, deprecated, message: "Use the asynchronous version of `disconnect` for increased safety")
     public func disconnect() {
-        clientUpdater.disconnect(source: .userInitiated) {
+        disconnect {}
+    }
+
+    /// Disconnects the chat client from the chat servers. No further updates from the servers
+    /// are received.
+    public func disconnect(completion: @escaping () -> Void) {
```
Comment (nuno-vieira) [inline Sources/StreamChat/ChatClient.swift:377]: I'm still concerned about this API change. Would be nice to have some customers try this new API and see the potential impact on their apps. If this is problematic we could opt in the approach of moving this to the login side.

Comment (polqf) [reply Sources/StreamChat/ChatClient.swift:377]: The previous API will still be available, under a deprecation note.

**Review by martinmitrevski — APPROVED** [review]
LGTM ✅

**Comment by nuno-vieira** [comment]
@polqf The Demo App looks that is not building

**Review by nuno-vieira — APPROVED** [review]

**Pull request merged**